### PR TITLE
[WIP] Stateful bots: improve test coverage.

### PR DESCRIPTION
**Pull request by neiljp**
**Current state**:
* Awaiting confirmation that changes to bot testing lib are good, and whether to switch to list[tuple]
* TicTacToe tests should be good
* VirtualFS tests are a preliminary change pending ordered changes, unless just using a loop like in tictactoe; further other work pending discussion on whether 'sample conversation' should be re-used between bot and tests, or just copied.
* Incrementor test is highly minimal pending adding update_message testing being possible, likely in another branch, perhaps after multi-response #40?

### CODE REVIEW

**Comment by timabbott** [comment]
This approach looks pretty good!  I hope we manage to have a gain finish with victory :)

**Comment by neiljp** [comment]
Thanks @timabbott. TicTacToe uses assert_bot_response, but since VirtualFS uses check_expected_responses I've added two extra commits to ensure that state isn't reused between sources (stream/private) and that the "expectations" are ordered, whereas the regular dict input isn't.

If we're open to tests being refactored, the List[Tuple[str,Any]] is more straightforward?

Also, particularly given the complexity of virtualfs and the 'sample conversation', I'm wondering if this can be split into a separate file used by both the bot and for testing?

**Comment by timabbott** [comment]
@neiljp can you rebase this and fix all the lint errors you'll get once the linter is running?

File: zulip_bots/zulip_bots/bots/virtual_fs/test_virtual_fs.py [hunk 1]
```diff
@@ -27,22 +24,25 @@ def test_bot(self):
...
         }
         self.check_expected_responses(expected)
+
+        expected = [
+            ("help", help_txt),
+            ("help ls", "foo_sender@zulip.com:\nsyntax: ls <optional_path>"),
+            ("", help_txt),
+            ("pwd", "foo_sender@zulip.com:\n/"),
+            ("cd /home", "foo_sender@zulip.com:\nERROR: invalid path"),
+            ("mkdir home", "foo_sender@zulip.com:\ndirectory created"),
```
Comment (timabbott) [inline zulip_bots/zulip_bots/bots/virtual_fs/test_virtual_fs.py]: can you add a final `cd /home` here?

**Comment by timabbott** [comment]
I merged the framework changes, after making a few tweaks to improve the variable names and make it all more readable.  

@neiljp can you rebase the actual tests and fix all the lint errors?  Apparently the linter wasn't running at whatever commit this is based on, and there's a ton of whitespace issues.

**Comment by timabbott** [comment]
Merged, thanks @neiljp!  I agree these are a lot harder to read than with the nicely whitespace-aligned code.  I'm just not sure what an easy way to quiet the PEP8 linter for just those blocks is.  May be worth googling a bit for suggestions, since it's a popular tool.

**Comment by neiljp** [comment]
It actually doesn't seem as bad now the comments are moved :)

**Pull request closed without merging**